File: src/frontend/nfe/lib/app/data/model/nfe_detalhe_imposto_ipi_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:nfe/app/infra/infra_imports.dart';
import 'package:nfe/app/data/domain/domain_imports.dart';

class NfeDetalheImpostoIpiModel {
	int? id;
	int? idNfeDetalhe;
	String? cnpjProdutor;
	String? codigoSeloIpi;
	int? quantidadeSeloIpi;
	String? enquadramentoLegalIpi;
	String? cstIpi;
	double? valorBaseCalculoIpi;
	double? quantidadeUnidadeTributavel;
	double? valorUnidadeTributavel;
	double? aliquotaIpi;
	double? valorIpi;

	NfeDetalheImpostoIpiModel({
		this.id,
		this.idNfeDetalhe,
		this.cnpjProdutor,
		this.codigoSeloIpi,
		this.quantidadeSeloIpi,
		this.enquadramentoLegalIpi,
		this.cstIpi,
		this.valorBaseCalculoIpi,
		this.quantidadeUnidadeTributavel,
		this.valorUnidadeTributavel,
		this.aliquotaIpi,
		this.valorIpi,
	});

	static List<String> dbColumns = <String>[
		'id',
		'cnpj_produtor',
		'codigo_selo_ipi',
		'quantidade_selo_ipi',
		'enquadramento_legal_ipi',
		'cst_ipi',
		'valor_base_calculo_ipi',
		'quantidade_unidade_tributavel',
		'valor_unidade_tributavel',
		'aliquota_ipi',
		'valor_ipi',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Cnpj Produtor',
		'Codigo Selo Ipi',
		'Quantidade Selo Ipi',
		'Enquadramento Legal Ipi',
		'Cst Ipi',
		'Valor Base Calculo Ipi',
		'Quantidade Unidade Tributavel',
		'Valor Unidade Tributavel',
		'Aliquota Ipi',
		'Valor Ipi',
	];

	NfeDetalheImpostoIpiModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idNfeDetalhe = jsonData['idNfeDetalhe'];
		cnpjProdutor = jsonData['cnpjProdutor'];
		codigoSeloIpi = jsonData['codigoSeloIpi'];
		quantidadeSeloIpi = jsonData['quantidadeSeloIpi'];
		enquadramentoLegalIpi = NfeDetalheImpostoIpiDomain.getEnquadramentoLegalIpi(jsonData['enquadramentoLegalIpi']);
		cstIpi = NfeDetalheImpostoIpiDomain.getCstIpi(jsonData['cstIpi']);
		valorBaseCalculoIpi = jsonData['valorBaseCalculoIpi']?.toDouble();
		quantidadeUnidadeTributavel = jsonData['quantidadeUnidadeTributavel']?.toDouble();
		valorUnidadeTributavel = jsonData['valorUnidadeTributavel']?.toDouble();
		aliquotaIpi = jsonData['aliquotaIpi']?.toDouble();
		valorIpi = jsonData['valorIpi']?.toDouble();
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idNfeDetalhe'] = idNfeDetalhe != 0 ? idNfeDetalhe : null;
		jsonData['cnpjProdutor'] = Util.removeMask(cnpjProdutor);
		jsonData['codigoSeloIpi'] = codigoSeloIpi;
		jsonData['quantidadeSeloIpi'] = quantidadeSeloIpi;
		jsonData['enquadramentoLegalIpi'] = NfeDetalheImpostoIpiDomain.setEnquadramentoLegalIpi(enquadramentoLegalIpi);
		jsonData['cstIpi'] = NfeDetalheImpostoIpiDomain.setCstIpi(cstIpi);
		jsonData['valorBaseCalculoIpi'] = valorBaseCalculoIpi;
		jsonData['quantidadeUnidadeTributavel'] = quantidadeUnidadeTributavel;
		jsonData['valorUnidadeTributavel'] = valorUnidadeTributavel;
		jsonData['aliquotaIpi'] = aliquotaIpi;
		jsonData['valorIpi'] = valorIpi;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idNfeDetalhe = plutoRow.cells['idNfeDetalhe']?.value;
		cnpjProdutor = plutoRow.cells['cnpjProdutor']?.value;
		codigoSeloIpi = plutoRow.cells['codigoSeloIpi']?.value;
		quantidadeSeloIpi = plutoRow.cells['quantidadeSeloIpi']?.value;
		enquadramentoLegalIpi = plutoRow.cells['enquadramentoLegalIpi']?.value != '' ? plutoRow.cells['enquadramentoLegalIpi']?.value : 'AAA';
		cstIpi = plutoRow.cells['cstIpi']?.value != '' ? plutoRow.cells['cstIpi']?.value : 'AAA';
		valorBaseCalculoIpi = plutoRow.cells['valorBaseCalculoIpi']?.value?.toDouble();
		quantidadeUnidadeTributavel = plutoRow.cells['quantidadeUnidadeTributavel']?.value?.toDouble();
		valorUnidadeTributavel = plutoRow.cells['valorUnidadeTributavel']?.value?.toDouble();
		aliquotaIpi = plutoRow.cells['aliquotaIpi']?.value?.toDouble();
		valorIpi = plutoRow.cells['valorIpi']?.value?.toDouble();
	}	

	NfeDetalheImpostoIpiModel clone() {
		return NfeDetalheImpostoIpiModel(
			id: id,
			idNfeDetalhe: idNfeDetalhe,
			cnpjProdutor: cnpjProdutor,
			codigoSeloIpi: codigoSeloIpi,
			quantidadeSeloIpi: quantidadeSeloIpi,
			enquadramentoLegalIpi: enquadramentoLegalIpi,
			cstIpi: cstIpi,
			valorBaseCalculoIpi: valorBaseCalculoIpi,
			quantidadeUnidadeTributavel: quantidadeUnidadeTributavel,
			valorUnidadeTributavel: valorUnidadeTributavel,
			aliquotaIpi: aliquotaIpi,
			valorIpi: valorIpi,
		);			
	}

	
}

File: src/backend/nfe/src/model/nfe_detalhe_imposto_ipi_model.py
from src import db


class NfeDetalheImpostoIpiModel(db.Model):
    __tablename__ = 'nfe_detalhe_imposto_ipi'

    id = db.Column(db.Integer, primary_key=True)
    cnpj_produtor = db.Column(db.String(14))
    codigo_selo_ipi = db.Column(db.String(60))
    quantidade_selo_ipi = db.Column(db.Integer)
    enquadramento_legal_ipi = db.Column(db.String(3))
    cst_ipi = db.Column(db.String(2))
    valor_base_calculo_ipi = db.Column(db.Float)
    quantidade_unidade_tributavel = db.Column(db.Float)
    valor_unidade_tributavel = db.Column(db.Float)
    aliquota_ipi = db.Column(db.Float)
    valor_ipi = db.Column(db.Float)
    id_nfe_detalhe = db.Column(db.Integer, db.ForeignKey('nfe_detalhe.id'))


    def mapping(self, data):
        self.id = data.get('id')
        self.id_nfe_detalhe = data.get('idNfeDetalhe')
        self.cnpj_produtor = data.get('cnpjProdutor')
        self.codigo_selo_ipi = data.get('codigoSeloIpi')
        self.quantidade_selo_ipi = data.get('quantidadeSeloIpi')
        self.enquadramento_legal_ipi = data.get('enquadramentoLegalIpi')
        self.cst_ipi = data.get('cstIpi')
        self.valor_base_calculo_ipi = data.get('valorBaseCalculoIpi')
        self.quantidade_unidade_tributavel = data.get('quantidadeUnidadeTributavel')
        self.valor_unidade_tributavel = data.get('valorUnidadeTributavel')
        self.aliquota_ipi = data.get('aliquotaIpi')
        self.valor_ipi = data.get('valorIpi')

    def serialize(self):
        return {
            'id': self.id,
            'idNfeDetalhe': self.id_nfe_detalhe,
            'cnpjProdutor': self.cnpj_produtor,
            'codigoSeloIpi': self.codigo_selo_ipi,
            'quantidadeSeloIpi': self.quantidade_selo_ipi,
            'enquadramentoLegalIpi': self.enquadramento_legal_ipi,
            'cstIpi': self.cst_ipi,
            'valorBaseCalculoIpi': self.valor_base_calculo_ipi,
            'quantidadeUnidadeTributavel': self.quantidade_unidade_tributavel,
            'valorUnidadeTributavel': self.valor_unidade_tributavel,
            'aliquotaIpi': self.aliquota_ipi,
            'valorIpi': self.valor_ipi,
        }